### PR TITLE
fix!: validate MsgTransfer before calling Transfer()

### PR DESCRIPTION
**Pull request by mpoke**
## Description

Closes: https://github.com/cosmos/interchain-security/issues/1237

Also, update the default consumer unbonding period to 2 weeks. 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [x] Added `!` to the type prefix if state-machine breaking change (i.e., requires coordinated upgrade)
* [x] Confirmed this PR does not introduce changes requiring state migrations, OR migration code has been added to consumer and/or provider modules
* [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [x] Provided a link to the relevant issue or specification
* [x] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules)
* [x] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [ ] Added a changelog entry to `CHANGELOG.md`
* [x] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] ~Updated the relevant documentation or specification~
* [x] Reviewed "Files changed" and left comments if necessary 
* [ ] Confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] confirmed this PR does not introduce changes requiring state migrations, OR confirmed migration code has been added to consumer and/or provider modules
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

### CODE REVIEW

File: tests/integration/distribution.go [hunk 1]
```diff
@@ -338,8 +340,156 @@ func (s *CCVTestSuite) TestEndBlockRD() {
 	}
 }
 
+// TestSendRewardsToProvider is effectively a unit test for TestSendRewardsToProvider(),
```
Comment (insumity) [inline tests/integration/distribution.go:343]: Maybe "a unit test for `SendRewardsToProvider`" instead of "a unit test for TestSendRewardsToProvider(),"?

File: tests/integration/distribution.go [hunk 2]
```diff
@@ -338,8 +340,156 @@ func (s *CCVTestSuite) TestEndBlockRD() {
...
+	testCases := []struct {
+		name           string
+		setup          func(sdk.Context, *consumerkeeper.Keeper, icstestingutils.TestBankKeeper)
+		expError       bool
+		tokenTransfers int
+	}{
+		{
+			name: "successful token transfer",
+			setup: func(ctx sdk.Context, keeper *consumerkeeper.Keeper, bankKeeper icstestingutils.TestBankKeeper) {
+				// setup transfer channel
```
Comment (insumity) [inline tests/integration/distribution.go:355]: `// setup transfer channel` comment here and below could probably be removed.

File: x/ccv/consumer/keeper/distribution.go [hunk 3]
```diff
@@ -103,62 +103,74 @@ func (k Keeper) shouldSendRewardsToProvider(ctx sdk.Context) bool {
...
-				if err != nil {
-					return err
-				}
-				sentCoins = sentCoins.Add(balance)
+	sourceChannelID := k.GetDistributionTransmissionChannel(ctx)
+	transferChannel, found := k.channelKeeper.GetChannel(ctx, transfertypes.PortID, sourceChannelID)
+	if !found || transferChannel.State != channeltypes.OPEN {
+		k.Logger(ctx).Info("WARNING: cannot send rewards to provider;",
+			"transmission channel not in OPEN state", "channelID", sourceChannelID)
+		return nil
```
Comment (insumity) [inline x/ccv/consumer/keeper/distribution.go:111]: Do we want to return an `err != nil` or maybe log an `Error` instead of `Info`. It seems that when we cannot [send the rewards](https://github.com/cosmos/interchain-security/blob/45c7e360fd8ffc3d312230822f514077705668f6/x/ccv/consumer/keeper/distribution.go#L34) we log a message with a level `Error`.

File: tests/integration/distribution.go [hunk 4]
```diff
@@ -338,8 +340,156 @@ func (s *CCVTestSuite) TestEndBlockRD() {
...
+					consumertypes.ConsumerToSendToProviderName,
+					sdk.NewCoins(sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(100))),
+				)
+				s.Require().NoError(err)
+			},
+			expError:       false,
+			tokenTransfers: 1,
+		},
+		{
+			name: "no transfer channel",
```
Comment (insumity) [inline tests/integration/distribution.go:376]: I guess this has to do with when `SendRewardsToProvider` returns an error but reading this test's name I would've expected to see `expError: true` because with "no transfer channel" we cannot really send rewards (?) Similarly for the tests below.

File: tests/integration/distribution.go [hunk 5]
```diff
@@ -338,8 +340,156 @@ func (s *CCVTestSuite) TestEndBlockRD() {
...
 }
 
+// TestSendRewardsToProvider is effectively a unit test for TestSendRewardsToProvider(),
+// but is written as an integration test to avoid excessive mocking.
+func (s *CCVTestSuite) TestSendRewardsToProvider() {
+	testCases := []struct {
+		name           string
+		setup          func(sdk.Context, *consumerkeeper.Keeper, icstestingutils.TestBankKeeper)
+		expError       bool
+		tokenTransfers int
```
Comment (insumity) [inline tests/integration/distribution.go:350]: Potentially this could be a `bool` because it would seem it only takes 2 values and then later on just check if `commitments` has one or zero elements?

**Comment by insumity** [comment]
Because this PR introduces `ValidateBasic` I would assume we would like to have a test with a bogus `MsgTransfer` and verify that the `SendRewardsToProvider` fails in those cases?
But it would seem we cannot easily do this because `MsgTransfer` is created inside the `SendRewardsToProvider` method?

File: tests/integration/distribution.go (hunk 5 above)
Comment (mpoke) [reply tests/integration/distribution.go:350]: Sure, but like this we avoid an extra if condition.

File: tests/integration/distribution.go (hunk 4 above)
Comment (mpoke) [reply tests/integration/distribution.go:376]: Yeah, but SendRewardsToProvider returns nil when there is no transfer channel. That's the reason we have `tokenTransfers`

File: x/ccv/consumer/keeper/distribution.go (hunk 3 above)
Comment (mpoke) [reply x/ccv/consumer/keeper/distribution.go:111]: This was the current behavior and I wouldn't want to change it as it works. I think it's fine to log a warning in case this happens. Since we don't return an error, it's weird to log an error.

**Comment by mpoke** [comment]
> Because this PR introduces `ValidateBasic` I would assume we would like to have a test with a bogus `MsgTransfer` and verify that the `SendRewardsToProvider` fails in those cases?
> But it would seem we cannot easily do this because `MsgTransfer` is created inside the `SendRewardsToProvider` method?

The input to this function is not the `MsgTrasfer`. As you already mentioned, the `MsgTrasfer` is created inside the function. `TestSendRewardsToProvider` already tests the `SendRewardsToProvider` function for different potential inputs. Is there another scenario we should test for?

**Comment by insumity** [comment]
> > Because this PR introduces `ValidateBasic` I would assume we would like to have a test with a bogus `MsgTransfer` and verify that the `SendRewardsToProvider` fails in those cases?
> > But it would seem we cannot easily do this because `MsgTransfer` is created inside the `SendRewardsToProvider` method?
> 
> The input to this function is not the `MsgTrasfer`. As you already mentioned, the `MsgTrasfer` is created inside the function. `TestSendRewardsToProvider` already tests the `SendRewardsToProvider` function for different potential inputs. Is there another scenario we should test for?

My understanding is that the only bogus `MsgTransfer` that is tested is for a "no recipient address," but `ValidateBasic` performs [additional checks](https://github.com/cosmos/ibc-go/blob/fe52d2a147592885c6aa189d9325b792c46c7b86/modules/apps/transfer/types/msgs.go#L88). I was  wondering if it makes sense to test for some of the other edge cases as well?

**Review by insumity — APPROVED** [review]
Thanks for your responses.

LGTM

**Review by p-offtermatt — APPROVED** [review]
LGTM. I find it a bit weird that things sometimes fail silently (i.e. we expect no error, but also expect no token transfers to be done), but doesn't seem too bad and also out of the scope of the PR

**Comment by mpoke** [comment]
> LGTM. I find it a bit weird that things sometimes fail silently (i.e. we expect no error, but also expect no token transfers to be done), but doesn't seem too bad and also out of the scope of the PR

The reason is that it's not really possible to handle an error in an endblocker, besides by panicking and halting the chain. In this case, when the transfer channel is not set, the rewards will accumulate. There is no reason to shut down.
